specs/core/0_beacon-chain.md

```diff
@@ -749,23 +758,32 @@ def add_validator(state: State,
                      sig=proof_of_possession,
                      domain=get_domain(state, current_slot, DOMAIN_DEPOSIT))
     # Pubkey uniqueness
-    assert pubkey not in [v.pubkey for v in state.validators]
-    rec = ValidatorRecord(
-        pubkey=pubkey,
-        withdrawal_credentials=withdrawal_credentials,
-        randao_commitment=randao_commitment,
-        randao_skips=0,
-        balance=DEPOSIT_SIZE * GWEI_PER_ETH,
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    if pubkey not in validator_pubkeys:
+        assert deposit_size == DEPOSIT_SIZE`
+        rec = ValidatorRecord(
+            pubkey=pubkey,
+            withdrawal_credentials=withdrawal_credentials,
+            randao_commitment=randao_commitment,
+            randao_skips=0,
+            balance=DEPOSIT_SIZE * GWEI_PER_ETH,
         status=status,
-        last_status_change_slot=current_slot,
-        exit_seq=0
-    )
-    index = min_empty_validator(state.validators)
-    if index is None:
-        state.validators.append(rec)
-        return len(state.validators) - 1
+            last_status_change_slot=current_slot,
+            exit_seq=0
+        )
+        index = min_empty_validator(state.validators)
+        if index is None:
+            state.validators.append(rec)
+            return len(state.validators) - 1
+        else:
+            state.validators[index] = rec
+            return index
     else:
-        state.validators[index] = rec
+        index = validator_pubkeys.index(pubkey)
+        val = state.validators[index]
+        assert val.withdrawal_credentials == withdrawal_credentials
+        assert deposit_size >= MIN_TOPUP_SIZE
```

Comment:
I suggest the deposits contraints (`MIN_DEPOSIT`, `MAX_DEPOSIT`) should go in the Vyper deposit contract.